src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,433 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+
+namespace Grpc.NetCore.HttpClient.Internal
+{
+    internal class GrpcCall<TRequest, TResponse>
+    {
+        private readonly CancellationTokenSource _callCts;
+        private readonly CancellationTokenRegistration? _ctsRegistration;
+        private readonly ISystemClock _clock;
+        private readonly TimeSpan? _timeout;
+        private readonly Timer _deadlineTimer;
+        private Metadata _trailers;
+        private CancellationTokenRegistration? _writerCtsRegistration;
+        private string _headerValidationError;
+
+        public bool DeadlineReached { get; private set; }
+        public bool Disposed { get; private set; }
+        public bool ResponseFinished { get; private set; }
+        public HttpResponseMessage HttpResponse { get; private set; }
+        public CallOptions Options { get; }
+        public Method<TRequest, TResponse> Method { get; }
+        public Task SendTask { get; private set; }
+        public HttpContentClientStreamWriter<TRequest, TResponse> ClientStreamWriter { get; private set; }
+        public HttpContentClientStreamReader<TRequest, TResponse> ClientStreamReader { get; private set; }
+
+        public GrpcCall(Method<TRequest, TResponse> method, CallOptions options, ISystemClock clock)
+        {
+            // Validate deadline before creating any objects that require cleanup
+            ValidateDeadline(options.Deadline);
+
+            _callCts = new CancellationTokenSource();
+            Method = method;
+            Options = options;
+            _clock = clock;
+
+            if (options.CancellationToken.CanBeCanceled)
+            {
+                // The cancellation token will cancel the call CTS
+                _ctsRegistration = options.CancellationToken.Register(CancelCall);
+            }
+
+            if (options.Deadline != null && options.Deadline != DateTime.MaxValue)
+            {
+                var timeout = options.Deadline.Value - _clock.UtcNow;
+                _timeout = (timeout > TimeSpan.Zero) ? timeout : TimeSpan.Zero;
+            }
+
+            if (_timeout != null)
+            {
+                // Deadline timer will cancel the call CTS
+                _deadlineTimer = new Timer(DeadlineExceeded, null, _timeout.Value, Timeout.InfiniteTimeSpan);
```

Comment:
What's the reason to use this instead of `.CancelAfter`?

Reply:
A different status is returned depending on whether the call was cancelled via a token or via the deadline being exceeded. A CTS doesn't provide a reason why it was cancelled (Cancel vs CancelAfter)

Reply:
It might be better to construct a new CTS with the specified timeout and register DeadlineExceeded with that instead of using a timer. That way when you dispose the registration, you wait for DeadlineExceeded to complete (if it's currently executing) automatically.